src/schematools/events/partial.py

```diff
@@ -0,0 +1,549 @@
+        table_id = self.schema_info.table_id
+        updates = {}
+        id_value = None if self.src_id is None else f"{self.src_id}.{self.src_volgnummer}"
+        if id_value is not None:
+            updates = {
+                "source_id": self.source_id,
+                f"{table_id}_id": id_value,
+            }
+        self._add_update(updates, "src_id", self.src_id, table_id)
+        self._add_update(updates, "src_volgnummer", self.src_volgnummer, table_id)
+        if self.schema_info.use_dimension_fields:
+            self._add_update(updates, "begin_geldigheid", self.begin_geldigheid)
+            self._add_update(updates, "eind_geldigheid", self.eind_geldigheid)
+        self.updates.update(updates)
+
+
+class EventsProcessor:
+    """The core event processing class. It needs to be initialised once
+    with configuration (datasets) and a db connection.
+    Once initialised, the process_event() method is able to
+    process incoming events.
+    The database actions are done using SQLAlchemy Core. So,
+    a helper function tables_factory() is used to created the
+    SA Tables that are needed during the processing of the events.
+    """
+
+    def __init__(
+        self,
+        datasets: List[DatasetSchema],
+        srid,
+        connection,
+        local_metadata=None,
+        truncate=False,
+    ):
+        self.datasets: Dict[str, DatasetSchema] = {ds.id: ds for ds in datasets}
+        self.srid = srid
+        self.conn = connection
+        _metadata = local_metadata or metadata  # mainly for testing
+        _metadata.bind = connection.engine
+        self.tables = {}
+        for dataset_id, dataset in self.datasets.items():
+            base_tables_ids = {dataset_table.id for dataset_table in dataset.tables}
+            self.tables[dataset_id] = tfac = tables_factory(dataset, _metadata)
+            self.geo_fields = defaultdict(lambda: defaultdict(list))
+            for table_id, table in tfac.items():
+                if not table.exists():
+                    table.create()
+                elif truncate:
+                    self.conn.execute(table.delete())
+                # self.has_compound_key = dataset_table.has_compound_key
+                # skip the generated nm tables
+                if table_id not in base_tables_ids:
+                    continue
+                for field in dataset.get_table_by_id(table_id).fields:
+                    if field.is_geo:
+                        self.geo_fields[dataset_id][table_id].append(field.name)
+
+    def process_relation(self, source_id, event_meta, event_data):
+        collection = event_meta["collection"]
+        schema_info = SchemaInfo.from_collection(collection, self.datasets)
+        event_type = event_meta["event_type"]
+        is_nm_table = schema_info.nm_table_id is not None
+        db_info = DbInfo.from_event_type(is_nm_table, event_type)
+
+        relation_handler = RelationHandler.fetch_handler(is_nm_table)(
```

Comment:
Waarom hier niet gewoon dit direct doen?
```suggestion
        relation_cls = M2MRelationHandler if is_nm_table else FKRelationHandler
        relation_handler = relation_cls(
```